File: clients/haskell/src/Model/DebugData.hs
module Model.DebugData where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.Vec2Double (Vec2Double)

-- | TODO - Document
data DebugDataCircle = DebugDataCircle {
    -- | TODO - Document
    pos :: Vec2Double,
    -- | TODO - Document
    radius :: Double }
    deriving Show

instance Trans DebugDataCircle where
    read = do
        pos <- Trans.read
        radius <- Trans.read
        return DebugDataCircle {
            pos,
            radius }
    
    write DebugDataCircle {
        pos,
        radius } = do
            Trans.write pos
            Trans.write radius

-- | TODO - Document
data DebugData
    -- | TODO - Document
    = Circle DebugDataCircle
    deriving Show

instance Trans DebugData where
    read = do
        tag :: Int32 <- Trans.read
        case tag of
            0 -> Circle <$> Trans.read
    
    write (Circle value) = do
        Trans.write (0 :: Int32)
        Trans.write value

File: clients/rust/src/model/debug_data.rs
use super::*;

/// TODO - Document
#[derive(Clone, Debug)]
pub enum DebugData {
    /// TODO - Document
    Circle {
        /// TODO - Document
        pos: model::Vec2F64,
        /// TODO - Document
        radius: f64,
    },
}

impl trans::Trans for DebugData {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Self::Circle {
                pos,
                radius,
            } => {
                <i32 as trans::Trans>::write_to(&0, writer)?;
                pos.write_to(writer)?;
                radius.write_to(writer)?;
            }
        }
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => {
                let pos: model::Vec2F64 = trans::Trans::read_from(reader)?;
                let radius: f64 = trans::Trans::read_from(reader)?;
                Ok(Self::Circle {
                    pos,
                    radius,
                })
            }
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}